[dbquery] Initial contribution

lujop opened the pull request:
# DBQuery Initial contribution

Initial contribution for a new addon that allows defining `Things` to execute native database queries.

This initial version provides support for InfluxDB 2.0 that is finished and usable.
In a future version in will I start to work after this, it will be submitted support for JDBC databases that currently OH support as persistence services.

A complete description of capabilities and usage of the addon is provided in it's README.

I created a new thread announcing the submission of this addon and for further discussion about it:
https://community.openhab.org/t/new-dbquery-addon-submitted/106640

The initial need for this addon was mainly discussed here:
https://community.openhab.org/t/influxdb-query-addon-proposal/95400/14

Here is another related issue that if sometime is implemented it will allow another DB implementation that uses an existing defined OH persistence service:
https://github.com/openhab/openhab-core/issues/1512

Skinah commented:
Fails to build due to not being 3.1.0-SNAPSHOT


[FATAL] Non-resolvable parent POM for org.openhab.addons.bundles:org.openhab.binding.dbquery:[unknown-version]: Could not find artifact org.openhab.addons.bundles:org.openhab.addons.reactor.bundles:pom:3.0.0-SNAPSHOT and 'parent.relativePath' points at wrong local POM @ line 7, column 11

lujop commented:
I've updated @Skinah and it's again ready to be reviewed.

lujop commented:
Hi @openhab/add-ons-maintainers 
I haven't seen any review activity in the PR. Is there any problem with it or is it just in the queue to be reviewed?

fwolter commented:
Frankly, I procrastinated the review of your PR, because I'm not confident if interfacing databases via a binding is the right approach.

lujop commented:
Thanks for being honest @fwolter.
If you have some time I will be grateful if you can review it, or start an argument discussion about why it can be accepted and where we can confront our opinions.
And if in the end it can't be accepted I prefer to know it and not be waiting indefinitely.

I developed it because it covered a real need for me, and in the process, I learned and dedicate considerable efforts trying to make it more idiomatic, flexible, and useful for the community. By now it has only the Influx2 implementation that is the one that I use. In fact, I stopped the development of the JDBC part because as I haven't any feedback I didn't want to develop something that I won't use if the community can benefit from it.

I suppose you're reticent about the addon being used as a substitute for the Openhab persistence API. But that's not the primary use of the addon.  And its configuration is totally independent of the persistence API. 
For example, I'm using it by now to integrate to an [IotaWatt device](https://iotawatt.com/) that stores electricity readings in an InfluxDB. 
And I'm sure that can be useful for other users to integrate to other things through a database.

If you want I can't remove the `Bypass limitations of current OpenHAb persistence queries` part from the documentation and add a disclaimer like 'Using this addon to query Openhab data it's unsupported, internal persistence format is considered an implementation detail and can be changed at any version'
Althought I've my opinion on that, and I think that giving a more low-level interface to uses can't be wrong, I see your point and I'm totally comfortable about not encouraging it's use from the addon.

I don't see that addon more harmful or wrong than it can be the exec or HTTP one that a lot of times is used to integrate things, not in the most ideal way.
Finally other smart home products have similar things like https://www.home-assistant.io/integrations/sql/

fwolter commented on bundles/org.openhab.binding.dbquery/README.md:
```diff
@@ -0,0 +1,214 @@
+# DBQuery Binding
+
+This binding allows creating items from the result of native database queries.
+It currently supports InfluxDB 2.0 and will support major relational databases through JDBC.
```
If you want to release it without JDBC support first, better don't mention it here.

lujop replied:
Ok, I will remove it. 
If I found some time, and now that you are reviewing it, and I hope it can be accepted after making desired corrections. I hope to work on the SQL binding soon but I think it's better to be able to integrate current work just In case I found little time and it comes later.

In fact, I want to try TimescaleDB for measurements and will need to implement that part because being able to integrate iotawatt readings is a must for me.

fwolter commented on bundles/org.openhab.binding.dbquery/README.md:
```diff
@@ -0,0 +1,214 @@
+be any supported database, and don't need to be the one you use as persistence service in InfluxDB.
+Some use cases can be:
+
+- Integrate a device that stores it's data in a database
```
```suggestion
- Integrate a device that stores its data in a database
```

fwolter commented on bundles/org.openhab.binding.dbquery/README.md:
```diff
@@ -0,0 +1,214 @@
+Some use cases can be:
+
+- Integrate a device that stores it's data in a database
+- Query derived data from you OpenHAB persistence, for example with Influx2 tasks you can process your data to create a new one   
```
```suggestion
- Query derived data from you openHAB persistence, for example with Influx2 tasks you can process your data to create a new one   
```

fwolter commented on bundles/org.openhab.binding.dbquery/README.md:
```diff
@@ -0,0 +1,214 @@
+
+## Supported Things
+
+There are two types of supported things: `DatabaseBridge` and a `Query`.
```
Thing type UIDs should start with lower case. Also, the real Thing Type UIDs should be mentioned here.
```suggestion
There are two types of supported things: `Influx2Bridge` and `Query`.
```

fwolter commented on bundles/org.openhab.binding.dbquery/README.md:
```diff
@@ -0,0 +1,214 @@
+ For InfluxDB use the `${paramName}` syntax for each parameter, and keep in mind that the values from that parameters must be from a trusted source as current
+ parameter substitution is subject to query injection attacks.
+ 
+#### timeout
+
+A time-out in seconds to wait for the query result, if it's exceeded, result will be discarded and the addon will do it's best effort to cancel the query.
+Currently it's ignored and it will be implemented in a future version.
```
If it's not implemented, better remove it from here.

fwolter commented on bundles/org.openhab.binding.dbquery/README.md:
```diff
@@ -0,0 +1,214 @@
+| resultContact   | Contact   | Result of last executed query as Contact, query must have `scalarResult=true` |
+| resultSwitch    | Switch    | Result of last executed query as Switch, query must have `scalarResult=true` |
+| parameters      | String    | Contains parameters of last executed query as JSON|
+| correct         | Switch    | True if the result of last executed query was correct |
```
```suggestion
| correct         | Switch    | ON if the result of last executed query was correct |
```

fwolter commented on bundles/org.openhab.binding.dbquery/pom.xml:
```diff
@@ -0,0 +1,107 @@
+  <dependencies>
+    <!-- influxdb-client-java -->
+    <dependency>
+      <groupId>com.influxdb</groupId>
```
Did you consider adding the dependencies via Karaf (feature.xml), so that the dependencies need to be downloaded only once if the influxdb persistence is also installed? (To make this work, the same adjustment needs to be done to the persistence addon)

lujop replied:
My intent, if it's possible, is to maintain independence between the persistence plugins.
To be able to evolve differently if needed and use different versions or implementations in the future if it's desired.

fwolter replied:
If the versions differ, both versions will be used. But if the versions are identical, the dependency will be reused. You won't introduce a dependency between the two.

lujop replied:
You're right.
Ok, I will try then but will need some support and guidance on that.
Is there any recent PR or documentation that I can take as a reference on how to do it?

Also, I will try to do all corrections in a row when you finish the review.

fwolter replied:
You could take a look here:  https://github.com/openhab/openhab-addons/blob/d36e1f10180f42ebe210f9178ab8af2a17a817b6/bundles/org.openhab.binding.ahawastecollection/src/main/feature/feature.xml#L7

lujop replied:
@fwolter: It's only adding that line and leaving the pom the same way?
Or it's needed to create the bundle too?

fwolter replied:
You don't need to create any bundle. But you need to change the scope in the pom.xml from compile to provided.

lujop replied:
If it's possible I prefer to leave this as it is for this first version.
Because I prefer to don't put more changes including modification of persistence addon.

fwolter reviewed (changes requested):
I reviewed the readme for now.

lujop commented:
Thank you very much @fwolter for taking your time to review.
I'm sorry to reply so many days later but I had very little time last week.

lujop commented:
Done suggested changes in https://github.com/openhab/openhab-addons/pull/8780/commits/4afe2afe92b85efe8ab0821f7cc78792347351d4

Skinah commented on bundles/org.openhab.binding.dbquery/README.md:
```diff
@@ -0,0 +1,213 @@
+This binding allows creating items from the result of native database queries.
+It currently only supports InfluxDB 2.0.
+
+You can use the addon in any situation where you want to create an item from a native query, the source of the query can
+be any supported database, and don't need to be the one you use as persistence service in openHAB.
```
```suggestion
You can use the addon in any situation where you want to create an item from a native query.
The source of the query can be any supported database, and doesn't need to be the one you use as the persistence service in openHAB.
```

Skinah commented on bundles/org.openhab.binding.dbquery/README.md:
```diff
@@ -0,0 +1,213 @@
+# DBQuery Binding
+
+This binding allows creating items from the result of native database queries.
+It currently only supports InfluxDB 2.0.
```
2.0 or 2.x ?

Skinah commented on bundles/org.openhab.binding.dbquery/README.md:
```diff
@@ -0,0 +1,213 @@
+- Query derived data from you openHAB persistence, for example with Influx2 tasks you can process your data to create a new one   
+- Bypass limitations of current openHAB persistence queries
+
+
```
```suggestion
```
I think the code analysis report will flag this. Only one empty line.

Skinah commented on bundles/org.openhab.binding.dbquery/README.md:
```diff
@@ -0,0 +1,213 @@
+## Supported Things
+
+There are two types of supported things: `influxdb2` and a `query`.
+For each different database, you want to connect you must define a `Bridge` thing for that database.
```
```suggestion
For each different database you want to connect to, you must define a `Bridge` thing for that database.
```

Skinah commented on bundles/org.openhab.binding.dbquery/README.md:
```diff
@@ -0,0 +1,213 @@
+
+There are two types of supported things: `influxdb2` and a `query`.
+For each different database, you want to connect you must define a `Bridge` thing for that database.
+Then each `Bridge` can define as much as wanted `Query` things you want to execute.
```
```suggestion
Then each `Bridge` can define as many `Query` things that you want to execute.
```

Skinah commented on bundles/org.openhab.binding.dbquery/README.md:
```diff
@@ -0,0 +1,213 @@
+For each different database, you want to connect you must define a `Bridge` thing for that database.
+Then each `Bridge` can define as much as wanted `Query` things you want to execute.
+
+Current supported `Bridge` is:
+
+- `influxdb2`
+
```
```suggestion
```
Not needed as the text above and below are enough.

Skinah commented on bundles/org.openhab.binding.dbquery/README.md:
```diff
@@ -0,0 +1,213 @@
+| scalarResult | false    | true     | If query always returns a single value or not                         |
+| scalarColumn | false    |          | In case of multiple columns, it indicates which to use for scalarResult|
+
+that are described further in the following subsections
```
```suggestion
These are described further in the following subsections
```

Skinah commented on bundles/org.openhab.binding.dbquery/README.md:
```diff
@@ -0,0 +1,213 @@
+ 
+#### timeout
+
+A time-out in seconds to wait for the query result, if it's exceeded, result will be discarded and the addon will do its best effort to cancel the query.
```
```suggestion
A time-out in seconds to wait for the query result, if it's exceeded, the result will be discarded and the addon will do its best to cancel the query.
```

Skinah commented on bundles/org.openhab.binding.dbquery/src/test/java/org/openhab/binding/dbquery/internal/Value2StateConverterTest.java:
```diff
@@ -0,0 +1,200 @@
+    }
+
+    private static Stream<Number> provideNumericTypes() {
+        return Stream.of(1L, 1.2, 1.2f, -1, 0, new BigDecimal("212321213123123123123123"));
```
How often does this run? Consider creating a static BigDecimal and reusing it. You are also using a similar line of code a new lines lower down in this same file. This appears to be parsing a String to create the BigDecimal when you can normally provide a number instead of a string to achieve the same thing.

lujop replied:
Done the static part, about the string it's used because I want to test a real big number that can't be expressed in a long literal string.
As this is a test I think that is best to test edge cases than premature optimization

Skinah commented on bundles/org.openhab.binding.dbquery/README.md:
```diff
@@ -0,0 +1,213 @@
+#### calculateParameters
+
+Triggers when there's a need to calculate parameters before query execution.
+When a query has `hasParameters=true` it fires the `calculateParameters` channel trigger and pause the execution until `setQueryParameters` action is call in
```
```suggestion
When a query has `hasParameters=true` it fires the `calculateParameters` channel trigger and pauses the execution until `setQueryParameters` action is call in
```

Skinah commented on bundles/org.openhab.binding.dbquery/README.md:
```diff
@@ -0,0 +1,213 @@
+When a query has `hasParameters=true` it fires the `calculateParameters` channel trigger and pause the execution until `setQueryParameters` action is call in
+ that query.
+ 
+In the case a query has parameters it's expected that there is some rule that catch the `calculateParameters` trigger, calculate the parameters
+ with the corresponding logic and then call the `setQueryParameters` action, after that the query will be really executed.
```
```suggestion
In the case a query has parameters, it's expected that there is a rule that catches the `calculateParameters` trigger, calculate the parameters
 with the corresponding logic and then calls the `setQueryParameters` action, after that the query will be executed.
```

Skinah commented on bundles/org.openhab.binding.dbquery/README.md:
```diff
@@ -0,0 +1,213 @@
+
+### The Simplest case 
+
+Define a InfluxDB2 database thing and a query with an interval execution. That executes the query every 15 seconds and punts the result in `myItem`
```
```suggestion
Define a InfluxDB2 database thing and a query with an interval execution.
That executes the query every 15 seconds and punts the result in `myItem`.
```
New line for each new sentence

Skinah commented on bundles/org.openhab.binding.dbquery/README.md:
```diff
@@ -0,0 +1,213 @@
+     log.info("execute query result is "+str(result.data))
+    
+
+Use this action with care, because as the query is executed synchronously is not good to execute long-running queries and block script execution.
```
```suggestion
Use this action with care, because as the query is executed synchronously, it is not good to execute long-running queries that can block script execution.
```

Skinah commented on bundles/org.openhab.binding.dbquery/README.md:
```diff
@@ -0,0 +1,213 @@
+
+#### setQueryParameters
+
+It's used for queries with parameters to set them. To execute the action you need to pass the parameters as a Map.
```
```suggestion
It's used for queries with parameters to set them.
To execute the action you need to pass the parameters as a Map.
```

Skinah commented on bundles/org.openhab.binding.dbquery/README.md:
```diff
@@ -0,0 +1,213 @@
+
+#### getLastQueryResult
+
+It can be used in scripts to get the last query result. It doesn't have any parameters and returns an `ActionQueryResult` as defined in `executeQuery` action.
```
```suggestion
It can be used in scripts to get the last query result.
It doesn't have any parameters and returns an `ActionQueryResult` as defined in `executeQuery` action.
```

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/java/org/openhab/binding/dbquery/action/DBQueryActions.java:
```diff
@@ -0,0 +1,125 @@
+    private @Nullable DatabaseBridgeHandler databaseBridgeHandler;
+
+    @Override
+    public void activate() {
+        logger.trace("activate actions");
+    }
+
+    @Override
+    public void deactivate() {
+        logger.trace("deactivate actions");
+    }
```
What is the point of these? They dont seem to serve a purpose as there is no code... See guidelines here on not using the logs instead of the debugger.

https://www.openhab.org/docs/developer/guidelines.html#f-logging

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/java/org/openhab/binding/dbquery/action/DBQueryActions.java:
```diff
@@ -0,0 +1,125 @@
+    @Override
+    @RuleAction(label = "Set query parameters", description = "Set query parameters for a query")
+    public void setQueryParameters(@ActionInput(name = "parameters") Map<String, @Nullable Object> parameters) {
+        logger.info("setQueryParameters {}", parameters);
```
info logging should not be used very often, suggest this should be debug or trace?

https://www.openhab.org/docs/developer/guidelines.html#f-logging

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/java/org/openhab/binding/dbquery/internal/DBQueryBindingConstants.java:
```diff
@@ -0,0 +1,53 @@
+    public static final String RESULT_DATETIME_CHANNEL_TYPE = "result-channel-datetime";
+    public static final String RESULT_CONTACT_CHANNEL_TYPE = "result-channel-contact";
+    public static final String RESULT_SWITCH_CHANNEL_TYPE = "result-channel-switch";
+
+    private DBQueryBindingConstants() {
+    }
```
```suggestion
```
It can be removed.

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/java/org/openhab/binding/dbquery/internal/DBQueryHandlerFactory.java:
```diff
@@ -0,0 +1,68 @@
+    protected @Nullable ThingHandler createHandler(Thing thing) {
+        ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+
+        logger.info("createHandler for thing {}", thing.getUID());
+
```
```suggestion
```
Info should be used very sparingly. The event log will tell you already when a thing changes to initializing, so this would be doubled up logging if you changed it to debug or trace. Hence why I suggest it is removed.

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/java/org/openhab/binding/dbquery/internal/DatabaseBridgeHandler.java:
```diff
@@ -0,0 +1,130 @@
+
+    @Override
+    public void initialize() {
+        logger.trace("initialize bridge {}", getThing().getUID());
```
```suggestion
```
IMHO should be removed as the event log already does this.

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/java/org/openhab/binding/dbquery/internal/DatabaseBridgeHandler.java:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    private void connectDatabase() {
+        logger.debug("connectDatabase {}", database);
```
IMHO this should be removed you don't need to log this, then log if it is successful or not successful as well.

lujop replied:
I do not completely agree with this for the following reasons:
- It doesn't do any warm to the user as to have it logged has to explicitly enable the debug level for the addon category.
- Connect to the database can come from initialization, retry attempts,...
- When a user reports a strange issue that is non easily reproducible, asking him to enable trace for the addon category is very useful as you can't debug his machine. I needed it in some difficult bugs in influxdb persistence addon and has been proven to be very useful.
- I see mainly all the persistence addons logging when connected, disconnected,...
- Finally, although is true that you can have the item status changelog, it's for all the items, isn't it? Having that info only for the addon, and with addon language, it's a valuable information

Skinah replied:
Leave as is then if it will help fault find for users, as changing it to trace does nothing as most people run at INFO level.

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/java/org/openhab/binding/dbquery/internal/DatabaseBridgeHandler.java:
```diff
@@ -0,0 +1,130 @@
+        updateStatus(ThingStatus.UNKNOWN);
+        completable.thenAccept(result -> {
+            if (result) {
+                logger.trace("Succesfully connected to database {}", getThing().getUID());
```
Does it only change to ONLINE if it connects? if so then remove this logging as it is redundant.

lujop replied:
This is one of the only logs that I left, I hope that it's ok because IMO establishing a connection is very relevant for a DB addon and only is trace level

Skinah replied:
Since it does not occur very often I don't see it as a big deal to leave as is. If it was once a second, then you would see savings on not having to constantly check if trace was on or off.

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/java/org/openhab/binding/dbquery/internal/DatabaseBridgeHandler.java:
```diff
@@ -0,0 +1,130 @@
+                logger.trace("Succesfully connected to database {}", getThing().getUID());
+                updateStatus(ThingStatus.ONLINE);
+            } else {
+                logger.trace("Connect to database {} failed", getThing().getUID());
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
```
```suggestion
                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
```
Dont log when you change thing status. The event log can supply a reason for the error. See
https://www.openhab.org/docs/developer/guidelines.html#f-logging

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/java/org/openhab/binding/dbquery/internal/DatabaseBridgeHandler.java:
```diff
@@ -0,0 +1,130 @@
+        completable.thenAccept(result -> {
+            Bridge bridge = getBridge();
+            if (result) {
+                logger.trace("Successfully disconnected to database {}", bridge != null ? bridge.getUID() : "null");
+                updateStatus(ThingStatus.OFFLINE);
+            } else {
+                logger.trace("Disconnect to database {} failed", bridge != null ? bridge.getUID() : "null");
+                updateStatus(ThingStatus.UNKNOWN, ThingStatusDetail.COMMUNICATION_ERROR);
```
Same as above. Provide clear and helpful reasons why the error or device went offline if possible and dont log it when the status changes.

lujop replied:
The same justification as above, I think that connection-disconnection are very important messages, and a very low-level debug category is used.
In case of problems, I think that will be very useful.

fwolter replied:
I understand your points, but keeping the logs clean from redundant messages is important. See https://next.openhab.org/docs/developer/guidelines.html#f-logging
```
void myFun() {
    logger.debug("And now the thing goes online"); // DONT, DONT, really DONT do that
    updateState(ThingState.ONLINE);
}
```

lujop replied:
I see your point and I will remove but I don't agree, I think that mainly all the persistence bindings had logs like this.
And the problem with the events.log is that if you do that way you can't have that relevant debug/trace logs only for an addon, can you?

For me is very useful to ask a user without a reproducible problem clear instructions, like set dbquery category log category to trace and send me the logs.

fwolter replied:
That's right, the events.log contains the thing status detail messages of all Things.

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/java/org/openhab/binding/dbquery/internal/JDBCBridgeHandler.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Override
+    public void initialize() {
+        // TODO: implement
```
```suggestion
```

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/java/org/openhab/binding/dbquery/internal/JDBCBridgeHandler.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        // TODO: implement
```
```suggestion
```

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/java/org/openhab/binding/dbquery/internal/QueryHandler.java:
```diff
@@ -0,0 +1,277 @@
+
+    @Override
+    public void initialize() {
+        logger.trace("initialize query handler for {}", getQueryIdentifier());
```
As above

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/java/org/openhab/binding/dbquery/internal/QueryHandler.java:
```diff
@@ -0,0 +1,277 @@
+
+    @Override
+    public void dispose() {
+        logger.trace("dispose query handler for {}", getQueryIdentifier());
```
See https://www.openhab.org/docs/developer/guidelines.html#f-logging

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/java/org/openhab/binding/dbquery/internal/QueryHandler.java:
```diff
@@ -0,0 +1,277 @@
+                executeQuery();
+            }
+        } else {
+            logger.debug("Query Thing can only handle RefreshType commands as the thing is read-only");
```
Should this be warn level?

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/java/org/openhab/binding/dbquery/internal/QueryHandler.java:
```diff
@@ -0,0 +1,277 @@
+
+    @Override
+    public void bridgeStatusChanged(ThingStatusInfo bridgeStatusInfo) {
+        logger.trace("bridgeStatusChanged to {} for {}", bridgeStatusInfo.getStatus(), getQueryIdentifier());
```
```suggestion
```
The event log does this.

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/java/org/openhab/binding/dbquery/internal/QueryHandler.java:
```diff
@@ -0,0 +1,277 @@
+
+    private final QueryExecution.QueryResultListener queryResultReceived = (QueryResult queryResult) -> {
+        synchronized (QueryHandler.this) {
+            logger.trace("queryResultReceived for {} : {}", getQueryIdentifier(), queryResult);
+            updateStateWithQueryResult(queryResult);
```
Does this change the state of the thing or channel and get logged in the event log? if so please remove the logger.trace line.

Skinah replied:
Leave as is.

Skinah commented:
OK I have gone through 35% of the LOC and it all looks pretty good, just mainly logging usage so far. Can you go over all uses of the logger and change them whilst you wait for someone to continue looking over the binding? Also look at the code report found in the target folder after you compile, any lines in there please fix...

I am also not sure I want to spend too much time reviewing a binding that may not get accepted? I looked over the threads and @kaikreuzer mentioned to get it discussed and voted on by the core maintainers. See here...
https://community.openhab.org/t/influxdb-query-addon-proposal/95400/12

So far I have not seen a definitive decision on which way this support should be added, either by binding or by another way. Until that is 100% clear I am not sure too many people are going to want to review this fully in depth.

Sounds like it is a good idea that people want, just which way to implement it is what needs to be decided on. Can we work on getting this answered so that this does not hang in limbo.

fwolter commented:
Thanks for your valuable review! I think @lujop binding's approach is ok. If the code itself is good, I would approve it.

lujop commented:
Thanks a lot for the review @Skinah .
I will follow your suggestions as soon as I can get some time to work on that.

Skinah commented:
No problem and sorry it has taken so long to get input/feedback. If you can close/resolve any of the comments when your happy, it will make it cleaner and let me know if you have questions on any remaining points. If you wish to apply any of the suggestions, make sure you put in the signoff as the very last line in your commit message or it will throw out the DCO signoff check. Since I have done mostly suggestions it will be quick if you accept and apply them. When I get some more spare time I'll continue on looking it over.

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/java/org/openhab/binding/dbquery/internal/QueryResultChannelUpdater.java:
```diff
@@ -0,0 +1,84 @@
+            Class<? extends State> targetType = calculateItemType(channel);
+            State state = value2StateConverter.convertValue(extractedResult, targetType);
+            channelStateUpdater.updateChannelState(channel, state);
+            logger.trace("Channel {} state updated to {}", channel.getUID(), state);
```
```suggestion
```
state changes already get logged to the event.log

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/java/org/openhab/binding/dbquery/internal/Value2StateConverter.java:
```diff
@@ -0,0 +1,139 @@
+            return ((Number) value).doubleValue() != 0d;
+        else if (value instanceof String) {
+            var svalue = (String) value;
+            return Boolean.parseBoolean(svalue) || (svalue.equalsIgnoreCase("on")) || svalue.equals("1");
```
The code analysis reports this as a WARNING.

svalue.equals("1")

should be "1".equals(svalue)
reason is only the value in the brackets is checked safely for NULL and wont throw a NPE. Please change all occurrences of this to remove the warnings.

Skinah replied:
ignore then as I have not compiled it, only guessed this would be an issue for the code analysis report that is found in the target folder. If it does not show up in the report then ignore.

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,125 @@
+
+	<channel-type id="execute-channel">
+		<item-type>Switch</item-type>
+		<label>Executed query channel</label>
```
```suggestion
		<label>Executed Query</label>
```
labels should be:
1. Short, less than 25 characters.
2. Each word starts in capitals unless it is a joining word like or, and, a ...
3. It will get used to auto name the items with when the framework does the feature "creates equipment from a thing".

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,125 @@
+	</channel-type>
+	<channel-type id="result-channel-string">
+		<item-type>String</item-type>
+		<label>Executed query result channel with String type</label>
```
```suggestion
		<label>Fetch String</label>
```

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,125 @@
+	</channel-type>
+	<channel-type id="result-channel-number">
+		<item-type>Number</item-type>
+		<label>Executed query result channel with Number type</label>
```
```suggestion
		<label>Number Result</label>
```

Skinah commented on bundles/org.openhab.binding.dbquery/README.md:
```diff
@@ -0,0 +1,213 @@
+| Channel Type ID | Item Type | Description                                                                                                                        |
+|-----------------|-----------|------------------------------------------------------------------------------------------------------------------------------------|
+| execute         | Switch    | Send `ON` to execute the query manually. It also indicates if query is currently running (`ON`) or not running (`OFF`)          |
+| resultString    | String    | Result of last executed query as an String |
```
```suggestion
| resultString    | String    | Result of last executed query as a String |
```

Skinah commented on bundles/org.openhab.binding.dbquery/README.md:
```diff
@@ -0,0 +1,213 @@
+| resultContact   | Contact   | Result of last executed query as Contact, query must have `scalarResult=true` |
+| resultSwitch    | Switch    | Result of last executed query as Switch, query must have `scalarResult=true` |
+| parameters      | String    | Contains parameters of last executed query as JSON|
+| correct         | Switch    | `ON` if the result of last executed query was correct |
```
Can you make this clearer as I am not sure how this is different to resultSwitch channel? Do you mean `correct`, or do you mean true like a boolean?

Skinah replied:
So do I understand it if I suggest this wording?
```
`ON` if the last executed query completed successfully, `OFF` if the query failed.
```
I feel the description does not explain what the channel is for when someone is new to the binding, perhaps I am a slow learner :)

lujop replied:
Thanks, sounds better

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,125 @@
+	</channel-type>
+	<channel-type id="result-channel-datetime">
+		<item-type>DateTime</item-type>
+		<label>Executed query result channel with DateTime type</label>
```
```suggestion
		<label>DateTime Result</label>
```

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,125 @@
+	</channel-type>
+	<channel-type id="parameters-channel">
+		<item-type>String</item-type>
+		<label>Last query parameters exposed as a JSON object</label>
```
```suggestion
		<label>JSON Result</label>
```

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,125 @@
+	</channel-type>
+	<channel-type id="result-channel-switch">
+		<item-type>Switch</item-type>
+		<label>Executed query result channel with Switch type</label>
```
```suggestion
		<label>Switch Result</label>
```

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,125 @@
+	</channel-type>
+	<channel-type id="result-channel-contact">
+		<item-type>DateTime</item-type>
+		<label>Executed query result channel with Contact type</label>
```
```suggestion
		<label>Contact Result</label>
```

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,125 @@
+	</channel-type>
+	<channel-type id="correct-channel">
+		<item-type>Switch</item-type>
+		<label>True if last query executed correctly, otherwise false</label>
```
Too long and not in Capitals, see above comments and examples.

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,125 @@
+
+		<config-description>
+			<parameter name="query" type="text" required="true">
+				<label>Query definition</label>
```
```suggestion
				<label>Query Definition</label>
```

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,125 @@
+				<context>script</context>
+			</parameter>
+			<parameter name="hasParameters" type="boolean">
+				<label>Query has parameters</label>
```
```suggestion
				<label>Query has Parameters</label>
```

lujop commented:
Thanks for the reviews @Skinah  and @fwolter and I'm sorry for the long delay correcting them.
I think that I addressed all the points by applying them or with comments.

I tried to comply the DCO by rebasing but with the main branch merge there were problems.
If it's ok when all will be ok, I will rebase the commits on top of the main branch without a merge and hard push it to the PR.

fwolter commented:
If you force-push your branch before merging into main, we can't see if something was (unintentionally) changed. It's ok if one commit contains the correct sign-off.

lujop commented:
> If you force-push your branch before merging into main, we can't see if something was (unintentionally) changed. It's ok if one commit contains the correct sign-off.

Then no problem I thought that DCO must pass. But if only one commit it's needed there's no problem, I think that all have it except the merge `main`  ones

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,125 @@
+	</channel-type>
+	<channel-type id="correct-channel">
+		<item-type>Switch</item-type>
+		<label>True if last query executed correctly, otherwise false</label>
```
```suggestion
		<label>Last Query Worked</label>
		<description>True if last query executed correctly</description>
```
Note the label is short (less then 25 characters) and each word starts with a captial. Description can be longer and give more detail.

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/resources/OH-INF/thing/jdbc-bridge.xml:
```diff
@@ -0,0 +1,36 @@
+				<description>Password to authenticate to the database</description>
+			</parameter>
+			<parameter name="type" type="text" required="true">
+				<label>Database type</label>
```
```suggestion
				<label>Database Type</label>
```

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,125 @@
+				<default>false</default>
+			</parameter>
+			<parameter name="scalarResult" type="boolean">
+				<label>Scalar result</label>
```
```suggestion
				<label>Scalar Result</label>
```

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,125 @@
+				<default>true</default>
+			</parameter>
+			<parameter name="scalarColumn" type="text" required="false">
+				<label>Scalar column name</label>
```
```suggestion
				<label>Scalar Column Name</label>
```

Skinah commented on bundles/org.openhab.binding.dbquery/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,125 @@
+				<label>Interval</label>
+				<description>
+					An interval, in seconds, the query will be repeatedly executed. Default values is 0, which means that
+					query is never executed manually (you need to send ON to execute channel)
```
```suggestion
					query is never executed automatically. You need to send the ON command each time you wish to execute.
```

Does this make more sense, or did I get it wrong?

lujop replied:
Yes it's absolutely correct.

lujop commented:
@fwolter @Skinah How do you see the PR to integrate?
If I can avoid moving dependencies to feature and postpone for a future PR I think that's all resolved.
I prefer to postpone because it implies modification in the other addon and I think that it will require some work for transitive dependencies.

I've also near to completion the JDBC part, but I prefer to use another PR to don't postpone this too much.
Does it seem ok for both of you?

fwolter commented on bundles/org.openhab.binding.dbquery/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,10 @@
+	xsi:schemaLocation="https://openhab.org/schemas/binding/v1.0.0 https://openhab.org/schemas/binding-1.0.0.xsd">
+
+	<name>DBQuery Binding</name>
+	<description>This is the binding for DBQuery.</description>
```
Could you be a bit more verbose, so that somebody who hasn't heard about it gets a clue what this is about?

fwolter commented on bundles/org.openhab.binding.dbquery/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,10 @@
+
+	<name>DBQuery Binding</name>
+	<description>This is the binding for DBQuery.</description>
+	<author>Joan Pujol Espinar</author>
```
The author tag is deprecated and should therefore be removed. See openhab/openhab-core#1844.

fwolter commented on bundles/org.openhab.binding.dbquery/src/main/resources/OH-INF/thing/jdbc-bridge.xml:
```diff
@@ -0,0 +1,36 @@
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+	<bridge-type id="jdbc">
```
If JDBC is not implemented, yet, this file should be removed. You can use `git stash` to save it for further versions.

fwolter commented on bundles/org.openhab.binding.dbquery/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,126 @@
+			<parameter name="query" type="text" required="true">
+				<label>Query Definition</label>
+				<description>Query definition using native query language</description>
+				<context>script</context>
```
Are you sure this context exists? It is not documented. See https://www.openhab.org/docs/developer/bindings/config-xml.html#xml-structure-for-configuration-descriptions

lujop replied:
Yes and it works:
https://community.openhab.org/t/is-it-possible-to-define-a-thing-parameter-to-be-multiline/102832/2

fwolter replied:
Cool, would be awesome if you could file a PR extending the documentation!

fwolter reviewed (changes requested):
You need to add your binding to `bom\openhab-addons\pom.xml`.

There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html` after building it with `mvn clean install`.

fwolter commented:
> I've also near to completion the JDBC part, but I prefer to use another PR to don't postpone this too much.

That's totally fine.

lujop commented on bundles/org.openhab.binding.dbquery/src/main/java/org/openhab/binding/dbquery/internal/DBQueryHandlerFactory.java:
```diff
@@ -0,0 +1,68 @@
+        } else if (THING_TYPE_JDBC_BRIDGE.equals(thingTypeUID)) {
+            return new JDBCBridgeHandler((Bridge) thing);
+        } else {
+            throw new UnnexpectedCondition("thingTypeUID " + thingTypeUID + " not expected");
```
But is that possible that the factory it's called for a binding it doesn't know about?
Usually, I prefer to program that defensively to allow fail-fast behavior

fwolter replied:
I think the factory can be called for a Thing type, which doesn't exist.

Fail-fast is good! In this case the framework takes care of it. Returning null is a common pattern in all bindings.

lujop commented on bundles/org.openhab.binding.dbquery/src/main/java/org/openhab/binding/dbquery/internal/DatabaseBridgeHandler.java:
```diff
@@ -0,0 +1,129 @@
+
+    protected void scheduleRetryConnectionAttempt() {
+        logger.trace("Scheduled retry connection attempt every {}", RETRY_CONNECTION_ATTEMPT_TIME_SECONDS);
+        retryConnectionAttemptFuture = Executors.newSingleThreadScheduledExecutor().scheduleWithFixedDelay(
+                this::connectDatabase, RETRY_CONNECTION_ATTEMPT_TIME_SECONDS, RETRY_CONNECTION_ATTEMPT_TIME_SECONDS,
+                TimeUnit.SECONDS);
```
Thanks @fwolter, I see a big mistake here on my part controlling the future but not the scheduler.
I didn't know about the framework schedulers, is there any documentation about what I can expect from them (sharing, min/max number of threads,...) and what are fair uses for them?
Are ok with large IO operations on edge cases like big connection timeouts?

fwolter replied:
The documentation doesn't answer these questions in detail. You may run network operations in the framework's scheduler. Many bindings use 30s for HTTP requests. But in general they should return fast. So, for example `Thread.sleep()` is a no go.

The number of threads in this pool can be configured by the user, but there's a reasonable default value (which I don't have in mind).

lujop commented on bundles/org.openhab.binding.dbquery/src/main/java/org/openhab/binding/dbquery/internal/QueryHandler.java:
```diff
@@ -0,0 +1,274 @@
+        int interval = config.getInterval();
+        if (interval != QueryConfiguration.NO_INTERVAL && scheduledQueryExecutionInterval == null) {
+            logger.trace("Scheduling query execution every {} seconds for {}", interval, getQueryIdentifier());
+            scheduledQueryExecutionInterval = Executors.newSingleThreadScheduledExecutor()
+                    .scheduleAtFixedRate(this::executeQuery, 0, interval, TimeUnit.SECONDS);
```
Ok on framework scheduler, but here, the scheduleAtFixedRate is "necessary" because it's the documented and intended behavior.
Possibly an improvement can be to allow the user to setup the fixedDelay behaviour in config.

fwolter replied:
If binding documentation talks about refresh intervals, they use `scheduleWithFixedDelay`. Usually you don't notice a difference and it makes it a lot easier for the scheduler.

lujop commented on bundles/org.openhab.binding.dbquery/src/main/java/org/openhab/binding/dbquery/internal/dbimpl/influx2/InfluxDBClientFacadeImpl.java:
```diff
@@ -0,0 +1,114 @@
+    public void query(String query, BiConsumer<Cancellable, FluxRecord> onNext, Consumer<? super Throwable> onError,
+            Runnable onComplete) {
+        var currentQueryAPI = queryAPI;
+        if (currentQueryAPI != null)
+            currentQueryAPI.query(query, onNext, onError, onComplete);
+        else
+            logger.warn("Query ignored as current queryAPI is null");
```
Added, but it seems that the formatter doesn't do that.
Is not the provided and enforced formatter the one that is invoked by `spotless apply` ?

fwolter replied:
Spotless doesn't add these braces, but the eclipse formatter do.

lujop commented on bundles/org.openhab.binding.dbquery/src/main/java/org/openhab/binding/dbquery/internal/dbimpl/mock/MockDatabase.java:
```diff
@@ -0,0 +1,110 @@
+ * @author Joan Pujol - Initial contribution
+ */
+@NonNullByDefault
+public class MockDatabase implements Database {
```
It was used by me as a neutral implementation.
I will check moving to test and implement a little test if I've time

fwolter replied:
Dead code shouldn't be part of the binding. You could use `git stash` to save it for future use.

lujop commented:
@fwolter @Skinah I've just summited another commit with your review changes, if I didn't forget something, I think that's all done.

Skinah commented:
There are a few comments not addressed yet like putting each sentence on its own line in the readme. Take care of them and lets get this to compile with a green tick, just triggered a rebuild in case it was the build systems fault it failed.

lujop commented:
> There are a few comments not addressed yet like putting each sentence on its own line in the readme. Take care of them and lets get this to compile with a green tick, just triggered a rebuild in case it was the build systems fault it failed.

Done, I think that now it's all addressed.
About the addon compilation seems to build correctly in local. And in Jenkins the error I think that is prior to addon compilation, isn't it?

Skinah commented:
Can you run a `mvn clean install` from the root folder NOT from inside `org.openhab.binding.dbquery/` folder? the latter folder does not check how it compiles as a whole project but is quicker to compile, so do it from the root folder and let it run. Then you will see what is failing. Also check what is in the code analysis report that is found in your target folder.

lujop commented:
> Can you run a `mvn clean install` from the root folder NOT from inside `org.openhab.binding.dbquery/` folder? the latter folder does not check how it compiles as a whole project but is quicker to compile, so do it from the root folder and let it run. Then you will see what is failing. Also check what is in the code analysis report that is found in your target folder.

Now seems to be ok, and I already checked the code analysis report and all seems ok to me

Skinah commented on bundles/org.openhab.binding.dbquery/pom.xml:
```diff
@@ -0,0 +1,107 @@
+      <groupId>com.squareup.okio</groupId>
+      <version>1.17.3</version>
+    </dependency>
+    <dependency>
+      <artifactId>commons-csv</artifactId>
+      <groupId>org.apache.commons</groupId>
+      <version>1.6</version>
+    </dependency>
```
Can you remove this library and use Java 11 methods instead? See https://github.com/openhab/openhab-addons/issues/7722

lujop replied:
@Skinah I'm sorry but it's not possible to do it in a safe way.
The addon doesn't use commons code at all, but the influxdb driver uses it and it's not documented as an optional dependency.
Also, I don't know if it applied to that dependency because up to my knowledge  there isn't any support for csv parsing in Java 11

Skinah replied:
@fwolter can you comment on this one please? thanks.

fwolter replied:
Seems to be a reasonable usage of Apache commons in this case.

Skinah commented:
I can't assist any further as the binding looks good to my noob eyes. Once @fwolter is happy I'm sure it will be good to merge as the build is showing as successful.

fwolter commented:
I just noticed, that the `NOTICE` file is missing. Can you please file another PR adding it?

lujop commented:
Thank you @fwolter and @Skinah  for your time reviewing the addon.

About the NOTICE, I will do the PR in the next days, also I will try to submit soon the JDBC part that it's quite ready and I'm already using in my installation
